enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,444 @@
+
+### Open Questions
+
+None that I can think of at the moment
+
+### Test Plan
+
+- Add unit tests to the Cluster Ingress Operator to make sure the
+IngressController resource is generated as exepected.
+
+- Add periodic nightly tests which install single-node in the cloud, add a few
+worker nodes to it, then run conformance tests to make sure we don't run into
+any problems not described in this enhancement.
+
+- Add periodic nightly tests which install a single-node "none"-platform cluster,
+add worker nodes to it, and check that ingress traffic still works as expected
+and recovers even after the `router-default` pod gets deleted and rescheduled.
+Make sure this is still true even after upgrades.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after the single control-plane node reboots by
+running conformance tests post-reboot.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after an upgrade by running conformance tests
+post-upgrade.
+
+TODO: Describe day-1-workers tests?
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+In the non-goals section it's mentioned that this enhancement does not apply to
+clusters which have been installed prior to the enhancement, so their upgrade
+is not discussed.
+
+This enhancement, to the best of my knowledge, should have no problems
+persisting across cluster upgrades. The Test Plan section describes how this
+will be tested.
+
+### Version Skew Strategy
+
+Does not apply, to the best of my understanding.
+
+### Operational Aspects of API Extensions
+
+- The "ControlPlane" IngressPlacement parameter value may not be used if the cluster's
+ControlPlaneTopology parameter is set to "External". 
+
+- This API change only affects the defaulting behavior of the IngressController CR,
+it does not add any new capabilities to OCP, or gives any more flexibility than there
+already was.
+
+- For administrators and support engineers, the IngressController is still the source
+of truth and where you need to look for understanding the router placement in practice.
+Nothing has changed in that regard.
+
+#### Failure Modes
+
+- The "ControlPlane" IngressPlacement parameter value may not be used if the cluster's
+ControlPlaneTopology parameter is set to "External". This should ideally be blocked by
+admission controllers for the Ingress config CR, if there are any. On top of that, the
+Ingress Operator should refuse to create an IngressController CR or reconcile an existing
+one while IngressPlacement is set to "ControlPlane" and the ControlPlaneTopology parameter
+is set to "External". This will harm cluster health and disable ingress traffic to the cluster.
```

Comment:
Can the installer's validation reject that combination?  I don't know if there's an admission plugin that prevents the user from updating these fields post-install.  

I don't think the ingress operator should refuse to create a default ingresscontroller in this scenario; the operator should try its hardest to make the cluster functional.  How about if the operator simply ignore `ingressPlacement` if `controlPlaneTopology` is "External"?

Reply:
The installer will not create an Infrastructure CR with the `controlPlaneTopology` set to "External". However, the installer does not validate modifications made by the user to manifests. There is the possibility that user (incorrectly) sets the `controlPlaneTopology`.

Reply:
>Can the installer's validation reject that combination?

So following what @staebler said, I guess the answer to that is no

>I don't know if there's an admission plugin that prevents the user from updating these fields post-install.

OK

>I don't think the ingress operator should refuse to create a default ingresscontroller in this scenario; the operator should try its hardest to make the cluster functional. How about if the operator simply ignore ingressPlacement if controlPlaneTopology is "External"?

Fair, I wasn't sure what's the attitude towards misconfigurations. So you're saying the behavior should be completely ignoring it (i.e., treating it as if it were Workers)?

Should the ingress operator log a warning? Set some .status on some CR? Or just flat out ignore it?

Reply:
And yes, I mean treat `ControlPlane`/`External` as if `ingressPlacement` were set to `Workers`.